clippy_lints/src/methods/owned_to_owned.rs

```diff
@@ -0,0 +1,43 @@
+use crate::utils::{is_copy, match_def_path, match_type, span_lint_and_sugg, sugg};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::ExprKind;
+use rustc_lint::LateContext;
+use rustc_middle::ty::TyS;
+
+use super::OWNED_TO_OWNED;
+
+pub fn lint(cx: &LateContext<'_>, expr: &hir::Expr<'_>, expected_path: &[&str], expected_type: Option<&[&str]>) {
+    if_chain! {
+        if let ExprKind::MethodCall(method_path, _, [arg], _) = &expr.kind;
+        if let Some(meth_did) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+        if let Some(snippet) = sugg::Sugg::hir_opt(cx, arg);
+        let return_type = cx.typeck_results().expr_ty(&expr);
+        let input_type = cx.typeck_results().expr_ty(arg);
+        if TyS::same_type(return_type, input_type);
+        if match_def_path(cx, meth_did, expected_path) &&
+            expected_type.map_or(true, |expected_type_path| match_type(cx,input_type,expected_type_path));
+        then {
+            if is_copy(cx,return_type) {
```

Comment:
Nothing seems to prevent `to_owned()` from being called on a custom type that implements `Copy` (like the `Kitten` type in my test), do we want this lint to fire on those cases?

Without this check, this will suggest replacing `to_owned` with `clone`, and then `clone_on_copy` will suggest dropping the `.clone()` entirely.

Reply:
Oh I see. It is *technically* possible but I think this situation is unusual enough that we shouldn't have a special case for it. The purpose of having a borrowed variant of a type is to avoid expensive copies. But if a type implements `Copy`, then it shouldn't be expensive to copy. In any case, I think the lint should just fire normally because `clone` is still better than `to_owned`.